### PR TITLE
Log subprocess input and output

### PR DESCRIPTION
**Pull request by tothtamas28**
This PR implements a helper function, `run_process_2`, that modifies the behavior of `run_process` in the following details:
* `run_process_2` does not stream to `stderr` (`write_stderr=False`) by default. `run_process` streams to `stderr` (`pipe_stderr=False`) by default.
* `run_process_2` does not allow the called process to inherit the parent's `stdin`.
* `run_process_2` logs `stdin`, `stderr`, `stdout` of the called process.
* `run_process_2` writes the PID with each log entry to correlate entries belonging to the same process.
* `run_process_2` logs if an exception is raised during the process (e.g. a `KeyboardInterrupt`).
* `run_process_2` has a more succinct and more structured log output.
* `run_process_2` always returns a `CompletedProcess` with `stdout` and `stderr` set. `run_process` only returns a non-`None` value if `PIPE` is passed for the corresponding argument.

These changes enable better integration of external tools into an application without compromising on debuggability (i.e. all information is logged but the caller's output streams are not polluted by the callee).

Example: [log.txt](https://github.com/user-attachments/files/16085849/log.txt)

### CODE REVIEW

File: pyk/src/pyk/kllvm/compiler.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ def compile_kllvm(target_dir: str | Path, *, verbose: bool = False) -> Path:
         args += ['--verbose']
 
     _LOGGER.info(f'Compiling pythonast extension: {module_file.name}')
-    run_process(args, logger=_LOGGER)
+    run_process_2(args, logger=_LOGGER)
```
Comment (tothtamas28) [inline pyk/src/pyk/kllvm/compiler.py:40]: Breaking change: `stderr` is not written.

File: pyk/src/pyk/kllvm/compiler.py [hunk 2]
```diff
@@ -87,7 +87,7 @@ def compile_runtime(
         args += ccopts
 
     _LOGGER.info(f'Compiling python extension: {module_file.name}')
-    run_process(args, logger=_LOGGER)
+    run_process_2(args, logger=_LOGGER)
```
Comment (tothtamas28) [inline pyk/src/pyk/kllvm/compiler.py:90]: Breaking change: `stderr` is not written.

File: pyk/src/pyk/kllvm/compiler.py [hunk 3]
```diff
@@ -123,7 +123,7 @@ def generate_hints(
 
     args = [str(interpreter), str(input_kore_file), '-1', str(hints_file), '--proof-output']
     _LOGGER.info(f'Generating hints: {hints_file.name}')
-    run_process(args, logger=_LOGGER)
+    run_process_2(args, logger=_LOGGER)
```
Comment (tothtamas28) [inline pyk/src/pyk/kllvm/compiler.py:126]: Breaking change: `stderr` is not written.

File: pyk/src/pyk/kllvm/load_static.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 
 def get_kllvm() -> Path:
     args = ['llvm-kompile', '--bindings-path']
-    proc = run_process(args, pipe_stdout=True)
+    proc = run_process_2(args)
```
Comment (tothtamas28) [inline pyk/src/pyk/kllvm/load_static.py:15]: Breaking change: `stderr` is not written.

File: pyk/src/pyk/kore/tools.py [hunk 5]
```diff
@@ -79,5 +79,5 @@ def _kore_print(
     if color is not None:
         args += ['--color', 'on' if color else 'off']
 
-    run_res = run_process(args, input=input)
+    run_res = run_process_2(args, input=input)
```
Comment (tothtamas28) [inline pyk/src/pyk/kore/tools.py:82]: Breaking change: `stderr` is not written.

File: pyk/src/pyk/kllvm/compiler.py (hunk 1 above)
Comment (tothtamas28) [reply pyk/src/pyk/kllvm/compiler.py:40]: New output:

```
In [0]: compile_kllvm('.', verbose=True)
INFO:pyk.kllvm.compiler:Compiling pythonast extension: _kllvm.cpython-310-x86_64-linux-gnu.so
INFO:pyk.kllvm.compiler:[PID=752090][exec] llvm-kompile pythonast --python /home/ttoth/.cache/pypoetry/virtualenvs/kframework-8Tiaz0lD-py3.10/bin/python --python-output-dir /home/ttoth/git/k/pyk --verbose
INFO:pyk.kllvm.compiler:[PID=752090][stde] +  /home/ttoth/git/k/k-distribution/bin/llvm-kompile-clang  python_ast nolto -fno-stack-protector -v /home/ttoth/git/k/k-distribution/target/release/k/lib/kllvm/python_src/ast.cpp --python /home/ttoth/.cache/pypoetry/virtualenvs/kframework-8Tiaz0lD-py3.10/bin/python -fuse-ld=lld --python-output-dir /home/ttoth/git/k/pyk
INFO:pyk.kllvm.compiler:[PID=752090][stde]   | +  /usr/bin/clang++-15 -Wno-override-module -Wno-return-type-c-linkage   /home/ttoth/git/k/k-distribution/bin/../lib/kllvm//libParser.a /home/ttoth/git/k/k-distribution/bin/../lib/kllvm//libAST.a /home/ttoth/git/k/k-distribution/bin/../lib/kllvm//libBinaryKore.a -Wl,-u,table_getArgumentSortsForTag -fPIC -shared -I/home/ttoth/git/k/k-distribution/bin/../include/ -fvisibility=hidden -I/home/ttoth/.pyenv/versions/3.10.13/include/python3.10 -I/home/ttoth/.cache/pypoetry/virtualenvs/kframework-8Tiaz0lD-py3.10/lib/python3.10/site-packages/pybind11/include -o /home/ttoth/git/k/pyk/_kllvm.cpython-310-x86_64-linux-gnu.so -ltinfo -lgmp -lgmpxx -lmpfr -lpthread -ldl -lffi -lunwind -I /home/ttoth/git/k/k-distribution/bin/../include/kllvm -I /home/ttoth/git/k/k-distribution/bin/../include --std=c++17 -fno-stack-protector /home/ttoth/git/k/k-distribution/target/release/k/lib/kllvm/python_src/ast.cpp -fuse-ld=lld
INFO:pyk.kllvm.compiler:[PID=752090][done] status=0 time=5.866s
Out[0]: PosixPath('/home/ttoth/git/k/pyk/_kllvm.cpython-310-x86_64-linux-gnu.so')
```

Comment (tothtamas28) [reply pyk/src/pyk/kllvm/compiler.py:40]: Old output:

```
In [0]: compile_kllvm('.', verbose=True)
INFO:pyk.kllvm.compiler:Compiling pythonast extension: _kllvm.cpython-310-x86_64-linux-gnu.so
INFO:pyk.kllvm.compiler:Running: llvm-kompile pythonast --python /home/ttoth/.cache/pypoetry/virtualenvs/kframework-8Tiaz0lD-py3.10/bin/python --python-output-dir /home/ttoth/git/k/pyk --verbose
+  /home/ttoth/git/k/k-distribution/bin/llvm-kompile-clang  python_ast nolto -fno-stack-protector -v /home/ttoth/git/k/k-distribution/target/release/k/lib/kllvm/python_src/ast.cpp --python /home/ttoth/.cache/pypoetry/virtualenvs/kframework-8Tiaz0lD-py3.10/bin/python -fuse-ld=lld --python-output-dir /home/ttoth/git/k/pyk
  | +  /usr/bin/clang++-15 -Wno-override-module -Wno-return-type-c-linkage   /home/ttoth/git/k/k-distribution/bin/../lib/kllvm//libParser.a /home/ttoth/git/k/k-distribution/bin/../lib/kllvm//libAST.a /home/ttoth/git/k/k-distribution/bin/../lib/kllvm//libBinaryKore.a -Wl,-u,table_getArgumentSortsForTag -fPIC -shared -I/home/ttoth/git/k/k-distribution/bin/../include/ -fvisibility=hidden -I/home/ttoth/.pyenv/versions/3.10.13/include/python3.10 -I/home/ttoth/.cache/pypoetry/virtualenvs/kframework-8Tiaz0lD-py3.10/lib/python3.10/site-packages/pybind11/include -o /home/ttoth/git/k/pyk/_kllvm.cpython-310-x86_64-linux-gnu.so -ltinfo -lgmp -lgmpxx -lmpfr -lpthread -ldl -lffi -lunwind -I /home/ttoth/git/k/k-distribution/bin/../include/kllvm -I /home/ttoth/git/k/k-distribution/bin/../include --std=c++17 -fno-stack-protector /home/ttoth/git/k/k-distribution/target/release/k/lib/kllvm/python_src/ast.cpp -fuse-ld=lld
INFO:pyk.kllvm.compiler:Completed in 5.706s with status 0: llvm-kompile pythonast --python /home/ttoth/.cache/pypoetry/virtualenvs/kframework-8Tiaz0lD-py3.10/bin/python --python-output-dir /home/ttoth/git/k/pyk --verbose
Out[0]: PosixPath('/home/ttoth/git/k/pyk/_kllvm.cpython-310-x86_64-linux-gnu.so')
```

**Comment by tothtamas28** [comment]
@dwightguth, this PR introduces some breaking changes in the `pyk.kllvm` module (listed above), please check if they're acceptable for the work you do. If not, I'll revert them.

**Comment by Baltoli** [comment]
I presume we'll want to eventually deprecate the old `run_process` and do a rename, once clients are no longer using it?

**Review by Baltoli — APPROVED** [review]

**Comment by tothtamas28** [comment]
> I presume we'll want to eventually deprecate the old `run_process` and do a rename, once clients are no longer using it?

Exactly.

**Review by gtrepta — APPROVED** [review]

**Pull request merged**